lib/elixir/test/elixir/task_test.exs

```diff
@@ -449,9 +449,103 @@ defmodule TaskTest do
       assert Task.shutdown(task, 1) == {:exit, :killed}
     end
 
-    test "shutdown/2 returns {:exit, :noproc} if task handled" do
+    test "returns {:exit, :noproc} if task handled" do
       task = create_dummy_task(:noproc)
       assert Task.shutdown(task, :brutal_kill) == {:exit, :noproc}
     end
   end
+
+  describe "pmap/2" do
+    test "timeout" do
+      assert catch_exit([:infinity] |> Task.pmap(&sleep/1, [timeout: 0]) |> Enum.to_list) ==
+             {:timeout, {Task.Supervised, :pmap, [0]}}
+    end
+  end
+
+  for {desc, concurrency} <- ["==": 4, "<": 2, ">": 8] do
+    describe "pmap/2 with max_concurrency #{desc} tasks" do
+      @opts [max_concurrency: concurrency]
+
+      test "maps an enumerable with fun" do
+        assert 1..4 |> Task.pmap(&sleep/1, @opts) |> Enum.to_list ==
+               [ok: 1, ok: 2, ok: 3, ok: 4]
+      end
+
+      test "maps an enumerable with mfa" do
+        assert 1..4 |> Task.pmap(__MODULE__, :sleep, [], @opts) |> Enum.to_list ==
+               [ok: 1, ok: 2, ok: 3, ok: 4]
+      end
+
+      test "maps an enumerable without leaking tasks" do
+        assert 1..4 |> Task.pmap(&sleep/1, @opts) |> Enum.to_list ==
+               [ok: 1, ok: 2, ok: 3, ok: 4]
+        refute_received _
+      end
+
+      test "returns errors for failures" do
+        Process.flag(:trap_exit, true)
+        assert 101..104 |> Task.pmap(&Process.sleep(&1) && exit(&1), @opts) |> Enum.to_list ==
+               [exit: 101, exit: 102, exit: 103, exit: 104]
+      end
+
+      test "shuts down unused tasks" do
+        assert [0, :infinity, :infinity, :infinity] |> Task.pmap(&sleep/1, @opts) |> Enum.take(1) ==
+               [ok: 0]
+        assert Process.info(self(), :links) == {:links, []}
+      end
+
+      test "shuts down unused tasks without leaking messages" do
+        assert [0, :infinity, :infinity, :infinity] |> Task.pmap(&sleep/1, @opts) |> Enum.take(1) ==
+               [ok: 0]
+        refute_received _
+      end
+
+      test "is zippable" do
+        task = 1..4 |> Task.pmap(&sleep/1, @opts) |> Stream.map(&elem(&1, 1))
+        assert Enum.zip(task, task) ==
+               [{1, 1}, {2, 2}, {3, 3}, {4, 4}]
+      end
+
+      test "with inner halt" do
+        assert 1..8 |> Enum.take(4) |> Task.pmap(&sleep/1, @opts) |> Enum.to_list ==
```

Comment:
Shouldn't this be `Stream.take(4)`. Otherwise I don't see a difference between this and starting with `1..4`